test/gtest/ucp/test_ucp_wireup.cc

```diff
@@ -950,3 +952,151 @@ UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_wireup_fallback,
 /* Test all available ib transports */
 UCP_INSTANTIATE_TEST_CASE_TLS(test_ucp_wireup_fallback,
                               ib, "ib")
+
+
+class test_ucp_wireup_fallback_amo : public test_ucp_wireup {
+    void init() {
+        size_t device_atomics_cnt = 0;
+
+        test_ucp_wireup::init();
+        sender().connect(&receiver(), get_ep_params());
+        for (ucp_lane_index_t lane = 0;
+             lane < ucp_ep_num_lanes(sender().ep()); lane++) {
+            uct_iface_attr_t iface_attr;
+
+            if (!get_iface_attr(sender().ep(), lane, &iface_attr)) {
+                continue;
+            }
+
+            if (iface_attr.cap.flags & UCT_IFACE_FLAG_ATOMIC_DEVICE) {
+                device_atomics_cnt++;
+            }
+        }
+        test_ucp_wireup::cleanup();
+
+        if (!device_atomics_cnt) {
+            UCS_TEST_SKIP_R("there are no TLs that support device atomics");
+        }
+    }
+
+    void cleanup() {
+        /* do nothing */
+    }
+
+protected:
+
+    bool use_device_amo(ucp_ep_h ep) {
+        ucp_ep_config_t *ep_config = ucp_ep_config(ep);
+
+        for (ucp_lane_index_t lane = 0; lane < UCP_MAX_LANES; ++lane) {
+            if (ep_config->key.amo_lanes[lane] != UCP_NULL_LANE) {
+                uct_iface_attr_t *iface_attr =
+                    ucp_worker_iface_get_attr(ep->worker,
+                                              ep_config->key.lanes[lane].rsc_index);
+
+                return (iface_attr->cap.flags & UCT_IFACE_FLAG_ATOMIC_DEVICE);
+            }
+        }
+
+        return false;
+    }
+
+    bool get_iface_attr(ucp_ep_h ep, ucp_lane_index_t lane,
+                        uct_iface_attr_t *iface_attr) {
+        uct_ep_h uct_ep = ep->uct_eps[lane];
+        if (uct_ep == NULL) {
+            return false;
+        }
+
+        if (ucp_wireup_ep_test(uct_ep) || ucp_proxy_ep_test(uct_ep)) {
+            ucp_proxy_ep_t *proxy_ep = ucs_derived_of(uct_ep,
+                                                      ucp_proxy_ep_t);
+            uct_ep = proxy_ep->uct_ep;
+        }
+
```

Comment:
you can use `ucp_ep_get_iface_attr` for instance. No need to request attrs, they are stored in `worker->wiface[x].attrs`

Reply:
done